### PR TITLE
csr implemented, needs testing

### PR DESCRIPTION
**Pull request by nganjehloo**

### CODE REVIEW

File: src/main/scala/components/control.scala [hunk 1]
```diff
@@ -39,36 +40,39 @@ class Control extends Module {
 
   val signals =
     ListLookup(io.opcode,
-      /*default*/           List(false.B, false.B, 3.U,   false.B, false.B,  false.B, false.B,    0.U,    0.U),
-      Array(                 /*  branch,  memread, toreg, add,     memwrite, immediate, regwrite, alusrc1,  jump */
+      /*default*/           List(false.B, false.B, false.B, 4.U,   false.B, false.B,  false.B, false.B,    0.U,    0.U),
```
Comment (powerjg) [inline src/main/scala/components/control.scala:43]: What's up with all of these changes? I would have expected only a single line to change (and maybe an extra bit added to the others).

File: src/main/scala/components/csr.scala [hunk 2]
```diff
@@ -0,0 +1,506 @@
+/* Describes register file that maintains machine state */
+
+package dinocpu
+
+import chisel3._
+import collection.mutable.LinkedHashMap
+import chisel3.util._
+import chisel3.util.BitPat
+import chisel3.util.experimental.BoringUtils
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: Needed? It would be nice to not use this. So far, it's not been necessary.

File: src/main/scala/components/csr.scala [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+  val misaligned_fetch = 0x0
+  val fetch_access = 0x1
+  val illegal_instruction = 0x2
+  val breakpoint = 0x3
+  val misaligned_load = 0x4
+  val load_access = 0x5
+  val misaligned_store = 0x6
+  val store_access = 0x7
+  val machine_ecall = 0xb
+  val all = {
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: What is this used for?

File: src/main/scala/components/csr.scala [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+    res.toArray
+  }
+}
+
+object MCSRs {
+  //machine information registers
+  val mvendorid = 0xf11 //vendor id
+  val marchid = 0xf12 //architecture id
+  val mimpid = 0xf13 //implementation id
+  val mhartid = 0xf14 //hardware thread id
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:38]: Not 0? Or is this an offset into something?

File: src/main/scala/components/csr.scala [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+    res += misaligned_load
+    res += load_access
+    res += misaligned_store
+    res += store_access
+    res += machine_ecall
+    res.toArray
+  }
+}
+
+object MCSRs {
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:33]: Can you add some documentation here?

File: src/main/scala/components/csr.scala [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+  //DONT NEED
+  
+  //machine counter/timers
+  val mcycle = 0xb00 //machine cycle counter
+  val minstret = 0xb02 //machine instructions retured counter
+  val mcycleh = 0xb80
+  val minstreth = 0xb82
+  //performance counter setup
+  val mcounterinhibit = 0x320
+  val all = {
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: Is this used?

File: src/main/scala/components/csr.scala [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+  val mepc = 0x341 //machine exception program counter
+  val mcause = 0x342 //machine trap cause
+  val mtval = 0x343 //machine bad address or instruction
+  val mip = 0x344 //machine interrupt pending
+
+  //machine memory protection
+  //DONT NEED
+  
+  //machine counter/timers
+  val mcycle = 0xb00 //machine cycle counter
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:59]: What are these numbers?!?!

File: src/main/scala/components/csr.scala [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+  val offset = UInt(7.W)
+}
+
+class MTVec extends Bundle{
+  val base = UInt(30.W)
+  val mode = UInt(2.W)
+}
+
+class MIx extends Bundle{
+ val wpri1 = UInt(20.W)
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:111]: Missing a space on these?

File: src/main/scala/components/csr.scala [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+    MCSRs.mtval -> reg_mtval,
+    MCSRs.mcause -> reg_mcause.asUInt(),
+    MCSRs.mhartid -> 0.U,
+    MCSRs.medeleg -> reg_medeleg)
+    
+  read_mapping += MCSRs.mcycleh -> 0.U 
+  read_mapping += MCSRs.minstreth -> 0.U
+  
+  //CSR DECODE
+  val cmd = if( io.decode.inst(6, 0) == ("b1110011".U) ) {
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: `if`s? I would have expected hardware here (e.g., when/elsewhen/otherwise or switch/case)

File: src/main/scala/components/helpers.scala [hunk 10]
```diff
@@ -4,7 +4,223 @@ package dinocpu
 
 import chisel3._
 import chisel3.util._
+import scala.math._
+import scala.collection.mutable.ArrayBuffer
 
+object Util
```
Comment (powerjg) [inline src/main/scala/components/helpers.scala]: Is this all needed? If so, I would make a new file called "util" instead of putting it here. 

I would rather not clutter the DINO CPU if we can avoid it.

File: src/test/resources/risc-v/sw.riscv [hunk 11]
```diff
@@ -2,6 +2,7 @@
   .align 2       # Make sure we're aligned to 4 bytes
   .globl _start
 _start:
+    addi t0, x0, 7
```
Comment (powerjg) [inline src/test/resources/risc-v/sw.riscv]: Why this change?

File: src/verilog/testbench.cpp [hunk 12]
```diff
@@ -0,0 +1,33 @@
+#include "VTop.h"
```
Comment (powerjg) [inline src/verilog/testbench.cpp]: Why this file added?

**Review by powerjg — COMMENTED** [review]
Looks like things are progressing!

The csr.scala file needs a lot more documentation. I got completely lost trying to understand it. What is the design, what are the interfaces, can we simplify at all since we're only implementing a subset?

File: src/main/scala/components/csr.scala (hunk 2 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: removed them except 1st 2 which are needed

File: src/main/scala/components/csr.scala (hunk 4 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:38]: index into the csr register file

File: src/main/scala/components/csr.scala (hunk 5 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:33]: Sure

File: src/main/scala/components/csr.scala (hunk 7 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:59]: index into the register file for that particular register

File: src/main/scala/components/helpers.scala (hunk 10 above)
Comment (nganjehloo) [reply src/main/scala/components/helpers.scala]: Will do. After some rewriting I'll see if I can get rid of that file

File: src/test/resources/risc-v/sw.riscv (hunk 11 above)
Comment (nganjehloo) [reply src/test/resources/risc-v/sw.riscv]: this was when I was testing store words in gem5 since I couldn't initialize the registers easily. I'll revert back now that I know stores work

File: src/verilog/testbench.cpp (hunk 12 above)
Comment (nganjehloo) [reply src/verilog/testbench.cpp]: When generating verilog from elaboration a testbench file is needed to run the verilated verilog. This file is just a simple testbench that clocks the pure verilated design. Should I .gitignore it?

File: src/main/scala/components/control.scala (hunk 1 above)
Comment (nganjehloo) [reply src/main/scala/components/control.scala:43]: Just to make sure again, this change should stay?

File: src/main/scala/components/csr.scala [hunk 13]
```diff
@@ -0,0 +1,533 @@
...
+  val reg_mscratch = Reg(UInt(32.W))
+  val reg_mtimecmp = Reg(UInt(64.W))
+  val reg_medeleg = Reg(UInt(32.W))
+
+  val reg_mip = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_mie = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_wfi = RegInit(false.B)
+  val reg_mtvec = RegInit(0.U.asTypeOf(new MTVec()))
+
+  val reg_time = WideCounter(64)
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:238]: Do we need to use wide counter? If so, can we just copy that into this file and get rid of `Util`?

File: src/main/scala/components/csr.scala [hunk 14]
```diff
@@ -0,0 +1,533 @@
...
+  val reg_instret = WideCounter(64, io.retire)
+
+  val reg_mcounterinhibit = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+  val reg_mcounteren = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+
+  val read_mstatus = io.status.asUInt()
+  val isa_string = "I"
+  val reg_misa = RegInit((BigInt(0) | isa_string.map(x => 1 << (x - 'A')).reduce(_|_)).U.asTypeOf(new MISA()))
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: Wow... the rocket chip people need to a little less crazy scala stuff. Could you at least write a comment on what this is doing? It might be better to just replace this with a hardcoded value (and a comment on what it means)

File: src/main/scala/components/csr.scala [hunk 15]
```diff
@@ -0,0 +1,533 @@
...
+  val reg_mvendorid = RegInit(0.U.asTypeOf(new MVendorID()))
+
+
+  val read_mapping = collection.mutable.LinkedHashMap[Int,Bits](
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:259]: A comment here would be useful. Especially since it took me so long to figure out :)

File: src/main/scala/components/csr.scala [hunk 16]
```diff
@@ -0,0 +1,533 @@
...
+    MCSRs.mip -> reg_mip.asUInt(),
+    MCSRs.mie -> reg_mie.asUInt(),
+    MCSRs.mscratch -> reg_mscratch,
+    MCSRs.mepc -> reg_mepc,
+    MCSRs.mtval -> reg_mtval,
+    MCSRs.mcause -> reg_mcause.asUInt(),
+    MCSRs.mhartid -> 0.U,
+    MCSRs.medeleg -> reg_medeleg)
+    
+  read_mapping += MCSRs.mcycleh -> 0.U
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: Move into the above map

File: src/main/scala/components/csr.scala [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+    cmd := MCSRCmd.nop
+    io.regwrite := false.B
+  }
+  
+  val csr = io.decode.inst(MCSRCmd.MSB, MCSRCmd.LSB)
+  val system_insn = cmd === MCSRCmd.interrupt
+  val cpu_ren = cmd =/= MCSRCmd.nop && !system_insn
+
+
+  val decoded_addr = read_mapping map { case (k, v) => k -> (csr === k) }
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: Add a comment here.

File: src/main/scala/components/csr.scala [hunk 18]
```diff
@@ -0,0 +1,533 @@
...
+  val priv_sufficient = MCSRCmd.MPRV >= csr(9,8)
+  val read_only = csr(11,10).andR
+  val cpu_wen = cpu_ren && cmd =/= MCSRCmd.read && priv_sufficient
+  val wen = cpu_wen && !read_only
+  val wdata = readModifyWriteCSR(cmd.asInstanceOf[UInt], io.rw.rdata, io.rw.wdata)
+
+  val opcode = 1.U << csr(2,0)
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:335]: This is weird code... Do you think it's a hardware optimization?

I was going to ask you to replace with a `when` statement, but now that I think about it maybe we shouldn't re-write too much until we have solid tests in place so we don't break things.

File: src/main/scala/components/csr.scala [hunk 19]
```diff
@@ -0,0 +1,533 @@
...
+
+  // ILLEGAL INSTR
+  when (io.exception) {
+    reg_mcause.interrupt := MCauses.illegal_instruction & "h80000000".U
+    reg_mcause.exceptioncode := MCauses.illegal_instruction & "h7fffffff".U
+    io.evec := "h80000004".U
+    reg_mepc := io.pc // misaligned memory exceptions not supported...
+  }
+
+  //assert(PopCount(insn_ret :: io.exception :: Nil) <= 1, "these conditions must be mutually exclusive")
```
Comment (powerjg) [inline src/main/scala/components/csr.scala:379]: Seems like a good check to have?

**Review by powerjg — COMMENTED** [review]
## git stuff
Do you mean to check in the verilog files? I doubt it, but I could be wrong.

For the binary files, I guess they are all different because you built them with a different version of the compiler? Either way, I don't think they should be checked in, unless something important changed.

BTW, I suggest when running `git add` you specify the exact files you want to add and avoid using `git add .`

## General comments
Overall, this looks really good, and now that it's getting closer to work (is working??) I took a close look at it. 

I have a few small things throughout. The only big comment I have is I'd like to see some comments/explanation in the csr file on what each register is. Specifically, in the csr register class where you're setting the values of different register, it would be useful to have some comments explaining the registers. For instance, `reg_misa` contains information on what RISC-V subset this CPU implements (32 vs 64, imadfvt....)

File: src/main/scala/single-cycle/cpu.scala [hunk 20]
```diff
@@ -68,18 +72,34 @@ class SingleCycleCPU(implicit val conf: CPUConfig) extends Module {
...
+  //MEMORY
   io.dmem.address   := alu.io.result
   io.dmem.writedata := registers.io.readdata2
   io.dmem.memread   := control.io.memread
   io.dmem.memwrite  := control.io.memwrite
   io.dmem.maskmode  := instruction(13,12)
   io.dmem.sext      := ~instruction(14)
 
+  //WRITEBACK
+  csr.io.decode.inst := instruction
```
[the remark this answers is not in the document]
Comment (nganjehloo) [reply src/main/scala/single-cycle/cpu.scala]: done

File: src/test/resources/risc-v/Makefile [hunk 21]
```diff
@@ -1,5 +1,5 @@
 
-RISCV ?= /opt/riscv
+RISCV ?= /home/nganjehl/riscv-gnu-toolchain
```
[the remark this answers is not in the document]
Comment (nganjehloo) [reply src/test/resources/risc-v/Makefile]: fixed

File: src/main/scala/components/csr.scala [hunk 22]
```diff
@@ -0,0 +1,533 @@
...
+  val evec = Output(UInt(32.W)) //
+  val exception = Input(Bool())  // rename to illgl inst
+  val retire = Input(Bool()) // rename to retire inst
+  val pc = Input(UInt(32.W)) //
+  val time = Output(UInt(32.W))//
+}
+
+class CSRRegFile extends Module{
+  //INIT CSR
+  val io = IO(new CSRRegFileIO)
```
[the remark this answers is not in the document]
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: fixed

File: src/main/scala/components/csr.scala [hunk 23]
```diff
@@ -0,0 +1,533 @@
...
+  io := DontCare
+
+  val reset_mstatus = WireInit(0.U.asTypeOf(new MStatus()))
+  reset_mstatus.mpp := MCSRCmd.MPRV//machine mode
+  reset_mstatus.mie := true.B//machine mode
+  reset_mstatus.mie := true.B//machine mode
```
[the remark this answers is not in the document]
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: accidentally pasted it twice. fixed

File: src/main/scala/components/csr.scala (hunk 14 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: fixed. hardcoded for I extension, but kept old code with comments just incase

File: src/main/scala/components/csr.scala (hunk 15 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:259]: done

File: src/main/scala/components/csr.scala (hunk 16 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: done

File: src/main/scala/components/csr.scala (hunk 17 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: done

File: src/main/scala/components/csr.scala (hunk 18 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:335]: Im not sure if it helps with optimization. I guess I'll leave it alone for now, but add a comment next to it

File: src/main/scala/components/csr.scala (hunk 19 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:379]: I was having an issue compiling with gem5 because of this line suddenly. It wanted a function called sc_time_stamp() to be implemented somewhere in the c++. I wasn't sure where to put it and wanted to make sure all the csr stuff was working before i went back to figure this out. It's a function that returns a double representing simulation time for systemC verilated code.

File: src/main/scala/components/csr.scala (hunk 13 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:238]: Since a lot of code in csr is based on coding style used in sodor removing Utils pretty much breaks everything. I was planning to remove it and refactor after I made sure everything with the CSR is done.

**Comment by powerjg** [comment]
Looking good! I'm excited things seem to be working!

Can you clean up this commit some? E.g., remove files that don't need to be updated, remove empty files, etc.?

**Comment by powerjg** [comment]
Oops... Sorry for the extra mail. Could you also rebase on top of current master so you can consume some of the fixes from Daniel? He's improved testing some (sorry, that'll be a pain to merge).

**Comment by nganjehloo** [comment]
I've gone ahead and rebased. this branch with master. I haven't merged with master.

File: src/test/scala/labs/Lab2Test.scala [hunk 24]
```diff
@@ -144,6 +144,7 @@ class SingleCycleStoreTesterLab2 extends CPUFlatSpec {
 */
 class SingleCycleLoadStoreTesterLab2 extends CPUFlatSpec {
 
+<<<<<<< HEAD
```
Comment (powerjg) [inline src/test/scala/labs/Lab2Test.scala]: Missed this in your rebase :)

File: src/test/scala/grading/Lab2Tests.scala [hunk 25]
```diff
@@ -440,21 +504,29 @@ class Lab2Grader extends JUnitSuite {
...
+                    Map("single-cycle" -> 1000, "five-cycle" -> 6, "pipelined" -> 6),
+                    Map(), 
+                    Map(),
+                    Map(5->23,6->20),
                     Map(5->23*20),
                     Map(), Map()),
         CPUTestCase("divider",
                     Map("single-cycle" -> 1000, "five-cycle" -> 6, "pipelined" -> 6),
+                    Map(), 
+                    Map(),
```
Comment (powerjg) [inline src/test/scala/grading/Lab2Tests.scala]: I notice that you never use this for checking or initializing. Is it necessary?

File: src/test/scala/grading/Lab2Tests.scala [hunk 26]
```diff
@@ -53,61 +53,85 @@ class Lab2Grader extends JUnitSuite {
...
         val tests = List[CPUTestCase](CPUTestCase("add1",
                       Map("single-cycle" -> 1),
+                      Map(), 
+                      Map(),
                       Map(5 -> 1234),
                       Map(0 -> 0, 5 -> 1234, 6 -> 1234),
                       Map(), Map()),
           CPUTestCase("addi1",
                       Map("single-cycle" -> 1, "five-cycle" -> 5, "pipelined" -> 5),
+                      Map(),
```
Comment (powerjg) [inline src/test/scala/grading/Lab2Tests.scala]: Hmm... I thought five-cycle was removed...

File: src/test/scala/labs/Lab2Test.scala (hunk 24 above)
Comment (nganjehloo) [reply src/test/scala/labs/Lab2Test.scala]: fixed

File: src/test/scala/grading/Lab2Tests.scala (hunk 26 above)
Comment (nganjehloo) [reply src/test/scala/grading/Lab2Tests.scala]: These files didn't pop up as conflicts during rebasing. Should I go ahead and remove the five cycle arguments manually? Master still has these as well. I'm guessing Daniel missed them?

File: src/test/scala/grading/Lab2Tests.scala (hunk 25 above)
Comment (nganjehloo) [reply src/test/scala/grading/Lab2Tests.scala]: I put it there as a place holder to initialize csr registers, but since I couldn't poke or peek because of the register names being optimized out, I left it there for the future when csr testing is resolved. Should I remove it?

Comment (powerjg) [reply src/test/scala/grading/Lab2Tests.scala]: Yeah... I know it was a pain to add... but since you're not using it now, it really should wait to be added until it's needed.

Comment (nganjehloo) [reply src/test/scala/grading/Lab2Tests.scala]: No problem. I'll go ahead and remove it.

File: src/main/scala/components/csr.scala (hunk 13 above)
Comment (powerjg) [reply src/main/scala/components/csr.scala:238]: Are you planning on merging this commit, then removing Util? or do you want to try that before merging?

File: src/main/scala/components/csr.scala (hunk 19 above)
Comment (powerjg) [reply src/main/scala/components/csr.scala:379]: ok.

File: src/main/scala/components/csr.scala [hunk 27]
```diff
@@ -0,0 +1,527 @@
...
+    //MCOUNTEREB IS FIXED IN THIS IMPLEMENTATION BECAUSE NO S | U MODE
+    
+    //MCOUNTINHIBIT
+    /* stops counting cycles and retired instructions if need be
+     *
+     */
+    when (decoded_addr(MCSRs.mcounterinhibit)) {
+      val new_mcounterinhibit = wdata.asTypeOf(new XCounterEnInhibit())
+      reg_mcounterinhibit := new_mcounterinhibit
+      if( reg_mcounterinhibit.cy == false.B) {
```
Comment (powerjg) [inline src/main/scala/components/csr.scala]: If not when?

File: src/main/scala/components/helpers.scala (hunk 10 above)
Comment (powerjg) [reply src/main/scala/components/helpers.scala]: Another bump. At a minimum remove what's not needed. I think `WideCounter` is the only thing I see in your code.

File: src/main/scala/testing/InstTests.scala [hunk 28]
```diff
@@ -30,71 +32,85 @@ object InstTests {
 	val rtype = List[CPUTestCase](
 		CPUTestCase("add1",
                 Map("single-cycle" -> 1, "pipelined" -> 5),
+								Map(), Map(),
```
Comment (powerjg) [inline src/main/scala/testing/InstTests.scala]: Did you mean to remove this?

**Comment by powerjg** [comment]
Also, check the travis to see why the build is failing.

File: src/main/scala/testing/InstTests.scala (hunk 28 above)
Comment (nganjehloo) [reply src/main/scala/testing/InstTests.scala]: yup. fixed. removed csr parameters from testing

File: src/main/scala/components/csr.scala (hunk 27 above)
Comment (nganjehloo) [reply src/main/scala/components/csr.scala]: oops lol. fixed

File: src/main/scala/components/csr.scala [hunk 29]
```diff
@@ -0,0 +1,527 @@
...
+  val SIZE = 3.W
+  val MSB = 31
+  val LSB = 20
+  val TRAPADDR = "h80000000".U
+  val MPRV = 3
+}
+
+class CSRRegFile extends Module{
+  //INIT CSR
+  val io = IO(new Bundle{
```
[the remark this answers is not in the document]
Comment (nganjehloo) [reply src/main/scala/components/csr.scala:186]: done

File: src/main/scala/components/helpers.scala (hunk 10 above)
Comment (nganjehloo) [reply src/main/scala/components/helpers.scala]: There were a lot of implicit functions I was using to avoid explicitly type casting everything, but I went ahead and removed all Util related code and changed csr to have WideCounter and just manually casted things as needed

**Comment by nganjehloo** [comment]
Looks like I'm getting noclassdefined error in my branch on travis, and the pull request version is failing the tests :(

--edit--
fixed pull request version. branch version is still getting noclassdeffound.

--edit2--
This is what is causing the noclassdef error

java.lang.NoClassDefFoundError: org/apache/commons/text/StringEscapeUtils

**Comment by powerjg** [comment]
I think that is #58. I'm not sure how to fix it. You may just be able to ask Travis to re run.

**Comment by DanG100** [comment]
I manually cleared the cache and reran the build. Also left instructions on how to do that in #58. Now that the branch are ahead of 169ced5f2c1705cf2a0df24a6f4d1ce99c84d23e it shouldn't happen again. If it does I'll looking the caching.

**Review by powerjg — APPROVED** [review]
Thanks for all of the updates! It looks like it's ready to go!

Optionally: It would be good to put a comment in the CSR file that much of the code comes from the rocket-chip repository or wherever you found the code. The rocket license should be compatible with BSD, so I think it's fine to copy paste the code.

**Comment by powerjg** [comment]
Make sure that when you commit this that the commit message makes sense.

**Comment by powerjg** [comment]
One more question: Do you have your tests integrated into the test suite? If not, that's something we should do in a follow up commit.

**Comment by nganjehloo** [comment]
Sure I'll add those comments in the CSR. Also I had csr tests defined in InstTest.scala and testing functions in CPUTestDriver, but because of the issues I was having in initializing the csr regs with treadle I opted to remove csr testing for now. Should I squash and merge or should we get testing working first?
